src/mbgl/text/quads.cpp

```diff
@@ -21,14 +21,19 @@ SymbolQuad getIconQuad(const Anchor& anchor,
                        const SymbolLayoutProperties::Evaluated& layout,
                        const float layoutTextSize,
                        const style::SymbolPlacementType placement, 
-                       const Shaping& shapedText) {
-    auto image = *shapedIcon.image();
+                       const Shaping& shapedText,
+                       const float pixelRatio) {
+    const SpriteAtlasElement& image = shapedIcon.image();
 
+    // TODO: all the border-based adjustments preserve existing behavior in rendering tests.
+    // However, I suspect the render tests themselves are incorrect, and the adjustments should
+    // all be removed.
     const float border = 1.0;
-    auto left = shapedIcon.left() - border;
-    auto right = left + image.pos.w / image.relativePixelRatio;
-    auto top = shapedIcon.top() - border;
-    auto bottom = top + image.pos.h / image.relativePixelRatio;
+
+    float left = shapedIcon.left() - border;
+    float right = left + image.size[0] + 2 * border * pixelRatio / image.pixelRatio;
+    float top = shapedIcon.top() - border;
+    float bottom = top + image.size[1] + 2 * border * pixelRatio / image.pixelRatio;
```

Comment:
Pretty sure now that this is indeed the cause of #6863, because the change only affects non-1x renders, and #6863 has been reported only on 2x renders.